### PR TITLE
Implement EIP-1153 for Cancun

### PR DESCRIPTION
**Pull request by gurukamath**
### What was wrong?
EIP-1153, which is slated for Cancun, is not currently implemented in the specs.

### How was it fixed?
Implemented EIP-1153

### CODE REVIEW

**Comment by gurukamath** [comment]
Rebased the EIP-1153 implementation on the `forks/cancun` branch

**Review by petertdavies — CHANGES_REQUESTED** [review]
I think the basic architecture of this PR is wrong.

Firstly `transient_state` is not a `State`. `State` comes with a lot of extra baggage (disk backing, state root calculating, etc) which doesn't apply to transient storage.

Secondly, transient storage is strictly a transaction scoped rather that block scoped. It is "discarded" at the end of the transaction. It belongs in the `Evm` dataclass with all of the other transaction scoped ephemeral data.

**Comment by gurukamath** [comment]
> I think the basic architecture of this PR is wrong.
> 
> Firstly `transient_state` is not a `State`. `State` comes with a lot of extra baggage (disk backing, state root calculating, etc) which doesn't apply to transient storage.
> 

You are right. Not sure how I missed  that. Will update

> Secondly, transient storage is strictly a transaction scoped rather that block scoped. It is "discarded" at the end of the transaction. It belongs in the `Evm` dataclass with all of the other transaction scoped ephemeral data.

The `Environment` data class is actually not entirely block scoped in the sense that a new `env` is created for each transaction within a block (`env` has to have `caller` and `origin` which vary by transaction). The `Evm` data class on the other hand has the scope of a message and might not be usable for multi-level inter-message call communication as is required by this EIP. I made this choice keeping those points in mind.

**Comment by gurukamath** [comment]
@petertdavies Updated the PR to include the `TransientStorage` class and include an attribute in the transaction scoped `EVM` class

File: src/ethereum/cancun/state.py [hunk 1]
```diff
@@ -603,3 +616,90 @@ def get_storage_original(state: State, address: Address, key: Bytes) -> U256:
     assert isinstance(original_value, U256)
 
     return original_value
+
+
+def backup_transient_storage(transient_storage: TransientStorage) -> None:
```
Comment (SamWilsn) [inline src/ethereum/cancun/state.py:621]: Does this function mark where a rollback will rollback to? Like this takes a snapshot? If so, I think this function's name should be something to do with "begin". It's pretty recognized as how you begin a transaction. We do this ourselves up in `begin_transaction`.

Comment (gurukamath) [reply src/ethereum/cancun/state.py]: Updated

File: src/ethereum/cancun/vm/interpreter.py [hunk 2]
```diff
@@ -257,6 +261,11 @@ def execute_code(message: Message, env: Environment) -> Evm:
     code = message.code
     valid_jump_destinations = get_valid_jump_destinations(code)
 
+    if message.parent_evm and message.parent_evm.transient_storage:
```
Comment (SamWilsn) [inline src/ethereum/cancun/vm/interpreter.py:264]: When would the parent not have a transient storage? Are there cases where an `Evm` would have a different `TransientStorage` than all of its parents?

Why not put the `TransientStorage` on the `Environment` instead?

Comment (gurukamath) [reply src/ethereum/cancun/vm/interpreter.py]: Updated the if condition. The `TransientStorage` was moved from the `Environment` to `Evm` data class after [this](https://github.com/ethereum/execution-specs/pull/850#pullrequestreview-1706392390) discussion.

**Review by petertdavies — APPROVED** [review]
Looks perfectly reasonable to me. We need to add support to the optimized state so the test suite passes.

**Pull request merged**